File: cmd/cozysoc/resolver_check.go
package main

import (
	"context"
	"errors"
	"flag"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/fijimunkii/cozysoc/internal/controller/api"
	"github.com/fijimunkii/cozysoc/internal/controller/localapi"
)

// This is presentation over the existing typed client, not a second executor.
// The terminal implementation is compiled; flags cannot replace it or consent.
type resolverCheckClient interface {
	CheckResolver(context.Context, string, func(context.Context, api.ResolverCheckReview) (bool, error)) (api.ResolverCheckResult, error)
}

func runResolverCheckCommand(ctx context.Context, args []string, stdout, stderr *os.File) (err error) {
	fs := flag.NewFlagSet("resolver-check", flag.ContinueOnError)
	fs.SetOutput(stderr)
	stateDir := fs.String("state-dir", "", "controller state directory")
	fs.Usage = func() {
		fmt.Fprintln(stderr, "Usage: cozysoc resolver-check [--state-dir PATH] SELECTION_ID\nExperimental macOS one-shot DNS/UDP check. Requires a terminal and an explicitly opted-in controller. Default: decline. No --yes, unattended mode or automatic retry.")
	}
	if err := fs.Parse(args); err != nil {
		if errors.Is(err, flag.ErrHelp) {
			return nil
		}
		return err
	}
	if fs.NArg() != 1 {
		return errors.New("resolver-check requires one opaque SELECTION_ID; flags must precede the target")
	}
	target := fs.Arg(0)
	if !localapi.ValidResolverSelectionID(target) {
		return errors.New("resolver-check requires a saved selection ID")
	}
	if err := ctx.Err(); err != nil {
		return err
	}
	// Refuse noninteractive use before loading controller credentials or dialing.
	terminal, err := openGatewayTerminal(os.Stdin, stdout)
	if err != nil {
		return errors.New("resolver-check requires normal-user foreground macOS terminal input and output; no controller connection was opened")
	}
	defer func() {
		if closeErr := terminal.Close(); closeErr != nil && err == nil {
			err = errors.New("terminal cleanup could not be confirmed; do not automatically retry a check")
		}
	}()
	dir, err := resolveStateDir(*stateDir)
	if err != nil {
		return err
	}
	return performResolverCheck(ctx, localapi.NewClient(dir), terminal, target)
}

func performResolverCheck(ctx context.Context, client resolverCheckClient, terminal gatewayCheckTerminal, target string) error {
	approved := false
	result, err := client.CheckResolver(ctx, target, func(decisionCtx context.Context, review api.ResolverCheckReview) (bool, error) {
		if err := gatewayPromptContext(decisionCtx); err != nil {
			return false, err
		}
		if err := terminal.Write(decisionCtx, resolverReviewText(review)); err != nil {
			return false, err
		}
		// Discard type-ahead, including incomplete canonical lines, AFTER the
		// disclosure and BEFORE showing the exact approval prompt. No raw mode.
		if err := terminal.FlushInput(); err != nil {
			return false, err
		}
		prompt := fmt.Sprintf("Type \"check %s\" to authorize ONLY this sample. Anything else declines.\nApproval (default: decline): ", review.SelectionID)
		if err := terminal.Write(decisionCtx, prompt); err != nil {
			return false, err
		}
		line, err := terminal.ReadLine(decisionCtx)
		if err != nil {
			return false, err
		}
		if err := gatewayPromptContext(decisionCtx); err != nil {
			return false, err
		}
		if line != "check "+review.SelectionID+"\n" {
			return false, nil
		}
		// An output failure must still fail before consent, not send invisibly.
		if err := terminal.Write(decisionCtx, "Requesting the reviewed one-shot sample; no automatic retry.\n"); err != nil {
			return false, err
		}
		if err := gatewayPromptContext(decisionCtx); err != nil {
			return false, err
		}
		approved = true
		return true, nil
	})
	if err != nil {
		return resolverCommandError(err, approved)
	}
	outputCtx, cancel := context.WithTimeout(ctx, 2*time.Second)
	defer cancel()
	if err := terminal.Write(outputCtx, resolverResultText(result)); err != nil {
		return resolverCommandError(err, approved)
	}
	if result.Outcome != "completed" && result.Outcome != "declined" {
		return errors.New("resolver run did not complete normally; see the measured evidence above; do not automatically retry")
	}
	return nil
}

// Input is validated by the typed client before it reaches the terminal.
func resolverReviewText(r api.ResolverCheckReview) string {
	p, b := r.Settings, r.Budget
	return fmt.Sprintf(`
EXPERIMENTAL: one-shot selected-resolver DNS/UDP check
Selection: %s
Resolver: %s (%s, %s)
Exact question: %s IN %s; expected: %s
Destination policy: %s; outside enrolled prefixes: %t
Source: %s
Interface: %s (index %d); enrolled routable prefixes: %s
Profile: %s
Created: %s
Review expires: %s (typing or resuming does not extend approval)

Limits: %d send call, %d DNS request bytes, %d reply bytes.
Receive ceiling: %d datagrams and %d calls; exchange %d ms; total %d ms.
Concurrency: %d run; at least %d ms between admissions per controller.
No TCP/EDNS fallback, referrals, host-resolver lookup, automatic retry or settings change.
DNS byte limits exclude IP/link overhead, neighbor resolution and server-side traffic.

Privacy: the exact query is visible to the resolver and network infrastructure.
The resolver MAY FORWARD IT UPSTREAM, even when its address is in enrolled prefixes.
Private settings remain in local SQLite. Run audits retain references, interface,
times and normalized DNS outcomes, not raw query names, replies or packets.
A response may be an error such as NXDOMAIN. A timeout does not prove an internet
outage or a security finding. This is one observer and one selected resolver.
Same-binding network reuse cannot be distinguished by interface/prefix matching.
Support is experimental native Terminal context; physical Wi-Fi/NICs, VPNs,
packaged Local Network permissions and sleep/resume are not certified.
Use only a network and query you may test.

`, r.SelectionID, p.Endpoint, p.Family, p.Transport, p.Name, p.QueryType, p.Expect, p.DestinationScope, r.OutsideEnrolledPrefixes, r.Source, r.Binding.InterfaceName, r.Binding.InterfaceIndex, strings.Join(r.Binding.Prefixes, ", "), r.Profile, r.CreatedAt.Format(time.RFC3339Nano), r.ExpiresAt.Format(time.RFC3339Nano), b.MaxSendCalls, b.MaxRequestBytes, b.MaxReplyBytes, b.MaxReceivedDatagrams, b.MaxReceiveCalls, b.ExchangeTimeoutMS, b.TotalTimeoutMS, b.MaxConcurrentRuns, b.MinRunIntervalMS)
}
func resolverResultText(r api.ResolverCheckResult) string {
	if r.Outcome == "declined" {
		return "Declined. No DNS request was authorized.\n"
	}
	text := fmt.Sprintf("Run: %s\nExecution outcome: %s\n", r.RunID, r.Outcome)
	if m := r.Measurement; m != nil {
		text += fmt.Sprintf("Historical sample: %s to %s\nRequest: %s; DNS exchange: %s\n", m.StartedAt.Format(time.RFC3339Nano), m.CompletedAt.Format(time.RFC3339Nano), m.Request, m.Exchange)
		if m.Reply != nil {
			text += fmt.Sprintf("Matched DNS response: RCODE %d; truncated: %t; answer classification: %s\n", m.Reply.RCode, m.Reply.Truncated, m.Reply.Answer)
		}
		if m.ResponseTimeNanoseconds != nil {
			text += fmt.Sprintf("Response time: %s (not necessarily successful resolution)\n", time.Duration(*m.ResponseTimeNanoseconds))
		}
	} else {
		text += "No measurement was confirmed.\n"
	}
	return text + "Historical evidence only. DNS errors/timeouts do not establish an internet outage or a security finding. Do not automatically retry.\n"
}
func resolverCommandError(err error, approved bool) error {
	if approved {
		return errors.New("resolver check outcome may be unknown; a request may have been sent; approval is consumed; do not automatically retry")
	}
	var response *localapi.ResponseError
	if errors.As(err, &response) && response.Code == "unavailable" {
		return errors.New("resolver checks require a separately started macOS controller with --experimental-resolver-checks and an active saved selection")
	}
	return errors.New("resolver review or terminal exchange failed; no approval was submitted; do not automatically retry")
}

File: ui/src/quality/resolver-check.ts
export interface ResolverSettings {
  endpoint: string;
  name: string;
  family: "ipv4" | "ipv6";
  transport: "udp";
  query_type: "A" | "AAAA";
  expect: "answer" | "nxdomain" | "no-data";
  destination_scope: "enrolled-prefix" | "exact-endpoint";
}
export interface ResolverSelection { selection_id: string; settings: ResolverSettings }
export interface ResolverReview extends ResolverSelection {
  review_id: string; created_at: string; expires_at: string; source: string;
  interface_name: string; interface_index: number; prefixes: string[];
  outside_enrolled_prefixes: boolean; may_forward_upstream: boolean;
  budget: { max_send_calls: number; max_request_bytes: number; max_reply_bytes: number;
    max_received_datagrams: number; max_receive_calls: number; exchange_timeout_ms: number;
    total_timeout_ms: number; max_concurrent_runs: number; min_run_interval_ms: number };
}
export interface ResolverCheckResult { outcome: "declined" | "completed" | "failed" | "canceled" | "indeterminate" | "blocked"; run_id?: string; failure_code?: string }
export interface ResolverCheckClient {
  reviewResolver(selectionID: string): Promise<ResolverReview>;
  decideResolver(reviewID: string, approve: boolean): Promise<ResolverCheckResult>;
}
const selectionIDPattern = /^selection\.[0-9a-f]{32}$/;
function invalid(): never { throw new Error("Resolver check response is invalid."); }
function exact(raw: unknown, required: string[], optional: string[] = []): Record<string, unknown> {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return invalid();
  const v = raw as Record<string, unknown>;
  if (required.some((k) => !Object.hasOwn(v, k)) || Object.keys(v).some((k) => !required.includes(k) && !optional.includes(k))) return invalid();
  return v;
}
function number(raw: unknown, low: number, high: number): number {
  if (typeof raw !== "number" || !Number.isSafeInteger(raw) || raw < low || raw > high) return invalid();
  return raw;
}
function time(raw: unknown): string {
  if (typeof raw !== "string" || !/^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(?:\.\d{1,9})?Z$/.test(raw) || !Number.isFinite(Date.parse(raw))) return invalid();
  return raw;
}
function settings(raw: unknown): ResolverSettings {
  const v = exact(raw, ["endpoint", "name", "family", "transport", "query_type", "expect", "destination_scope"]);
  if (typeof v.endpoint !== "string" || v.endpoint.length > 64 || !/^(?:\[[0-9a-f:]+\]|[0-9.]+):53$/.test(v.endpoint) ||
    typeof v.name !== "string" || v.name.length > 254 || !/^[A-Za-z0-9._-]+\.$/.test(v.name) ||
    (v.family !== "ipv4" && v.family !== "ipv6") || v.transport !== "udp" || (v.query_type !== "A" && v.query_type !== "AAAA") ||
    !["answer", "nxdomain", "no-data"].includes(String(v.expect)) || !["enrolled-prefix", "exact-endpoint"].includes(String(v.destination_scope))) return invalid();
  return v as unknown as ResolverSettings;
}
function selection(raw: unknown): ResolverSelection {
  const v = exact(raw, ["selection_id", "settings"]);
  if (typeof v.selection_id !== "string" || !selectionIDPattern.test(v.selection_id)) return invalid();
  return { selection_id: v.selection_id, settings: settings(v.settings) };
}
export function parseResolverSelections(raw: unknown): ResolverSelection[] {
  const v = exact(raw, ["items"]);
  if (!Array.isArray(v.items) || v.items.length > 16) return invalid();
  const items = v.items.map(selection);
  if (new Set(items.map((i) => i.selection_id)).size !== items.length) return invalid();
  return items;
}
export function parseResolverReview(raw: unknown): ResolverReview {
  const v = exact(raw, ["review_id", "selection_id", "settings", "created_at", "expires_at", "source", "interface_name", "interface_index", "prefixes", "outside_enrolled_prefixes", "may_forward_upstream", "budget"]);
  if (typeof v.review_id !== "string" || !/^[A-Za-z0-9_-]{43}$/.test(v.review_id) || typeof v.selection_id !== "string" || !selectionIDPattern.test(v.selection_id) ||
    typeof v.source !== "string" || v.source.length > 64 || !/^[0-9a-f:.]+$/.test(v.source) || typeof v.interface_name !== "string" || !/^[A-Za-z0-9][A-Za-z0-9_.:-]{0,63}$/.test(v.interface_name) ||
    !Array.isArray(v.prefixes) || v.prefixes.length < 1 || v.prefixes.length > 64 || v.prefixes.some((p) => typeof p !== "string" || p.length > 128 || !/^[0-9a-f:./]+$/.test(p)) ||
    typeof v.outside_enrolled_prefixes !== "boolean" || v.may_forward_upstream !== true) return invalid();
  const created = time(v.created_at), expires = time(v.expires_at);
  if (Date.parse(expires) - Date.parse(created) !== 30000) return invalid();
  const b = exact(v.budget, ["max_send_calls", "max_request_bytes", "max_reply_bytes", "max_received_datagrams", "max_receive_calls", "exchange_timeout_ms", "total_timeout_ms", "max_concurrent_runs", "min_run_interval_ms"]);
  const budget = { max_send_calls:number(b.max_send_calls,1,1), max_request_bytes:number(b.max_request_bytes,17,512), max_reply_bytes:number(b.max_reply_bytes,512,512),
    max_received_datagrams:number(b.max_received_datagrams,16,16), max_receive_calls:number(b.max_receive_calls,512,512), exchange_timeout_ms:number(b.exchange_timeout_ms,2000,2000),
    total_timeout_ms:number(b.total_timeout_ms,5000,5000), max_concurrent_runs:number(b.max_concurrent_runs,1,1), min_run_interval_ms:number(b.min_run_interval_ms,60000,60000) };
  return { review_id:v.review_id, selection_id:v.selection_id, settings:settings(v.settings), created_at:created, expires_at:expires,
    source:v.source, interface_name:v.interface_name, interface_index:number(v.interface_index,1,2147483647), prefixes:v.prefixes as string[],
    outside_enrolled_prefixes:v.outside_enrolled_prefixes, may_forward_upstream:true, budget };
}
export function parseResolverResult(raw: unknown): ResolverCheckResult {
  const v = exact(raw, ["outcome"], ["run_id", "failure_code"]);
  if (v.outcome === "declined") {
    if (v.run_id !== undefined || v.failure_code !== undefined) return invalid();
    return { outcome:"declined" };
  }
  if (!["completed", "failed", "canceled", "indeterminate", "blocked"].includes(String(v.outcome)) || typeof v.run_id !== "string" || !/^[0-9a-f]{32}$/.test(v.run_id)) return invalid();
  const expected: Record<string, string[]> = { completed:[""], failed:["execution_failed"], canceled:["canceled"], indeterminate:["execution_failed"], blocked:["precondition_failed", "review_expired"] };
  const failure = v.failure_code === undefined ? "" : v.failure_code;
  if (typeof failure !== "string" || !expected[String(v.outcome)]?.includes(failure)) return invalid();
  return { outcome:v.outcome as ResolverCheckResult["outcome"], run_id:v.run_id, ...(failure ? { failure_code:failure } : {}) };
}
